### PR TITLE
Threading read rows for CrossEntropy & LogLoss

### PR DESCRIPTION
**Pull request by masdevas**
Threading reading rows was added for LogLoss & CrossEntropy

### CODE REVIEW

**Comment by masdevas** [comment]
http://intel-ci.intel.com/eb3dde57-9776-f18e-a2b0-8cdcd4b723a1

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 1]
```diff
@@ -199,8 +199,9 @@ void addHessInPt(algorithmFPType * h, const algorithmFPType * xi, const algorith
 }
 
 template <typename algorithmFPType, Method method, CpuType cpu>
-services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute(const algorithmFPType * x, const algorithmFPType * y, size_t nRows,
-                                                                                 size_t n, size_t p, NumericTable * betaNT, NumericTable * valueNT,
+services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute(const NumericTable * dataNT,
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:202]: prefer reference over pointers when you don't need to check for null. Otherwise, should be checks for nullptr

```suggestion
services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute(const NumericTable & dataNT,
```

or 

```suggestion
services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute(const NumericTablePtr & dataNT,
```

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 2]
```diff
@@ -524,8 +534,11 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::doCompute
             const algorithmFPType interceptFactor = (parameter->interceptFlag ? 1 : 0);
             const auto hSize                      = nBeta * nBeta;
             TlsSum<algorithmFPType, cpu> tlsData(hSize);
-
+            SafeStatus safeStat;
             daal::threader_for(n, n, [&](size_t i) {
+                ReadRows<algorithmFPType, cpu> xr(const_cast<NumericTable *>(dataNT), i, 1);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:539]: A big degradation can happen here. Each line in parallel is converted, not processed by the block. SOA tables will be very expensive here.

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 3]
```diff
@@ -576,16 +589,14 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::compute(N
         TArrayScalable<algorithmFPType, cpu> aX(n * p);
         TArrayScalable<algorithmFPType, cpu> aY(n);
         s |= objective_function::internal::getXY<algorithmFPType, cpu>(dataNT, dependentVariablesNT, ntInd, aX.get(), aY.get(), nRows, n, p);
-        s |= doCompute(aX.get(), aY.get(), nRows, n, p, betaNT, valueNT, hessianNT, gradientNT, nonSmoothTermValue, proximalProjection,
-                       lipschitzConstant, parameter);
+        auto internalDataNT               = HomogenNumericTable<algorithmFPType>::create(aX.get(), p, n);
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:592]: Illegal instructions are expected here on old instruction sets. Use
```suggestion
        auto internalDataNT               = HomogenNumericTableCpu<algorithmFPType, cpu>::create(aX.get(), p, n);
```

Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:592]: Considering that it will work with a solver, and at each iteration, do you suggest allocating data? I would suggest cashing a block of allocated memory.

File: cpp/daal/src/algorithms/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i [hunk 4]
```diff
@@ -123,12 +123,13 @@ void LogLossKernel<algorithmFPType, method, cpu>::sigmoid(const algorithmFPType
 }
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/logistic_loss/logistic_loss_dense_default_batch_impl.i:123]: The same comments as above.

**Review by PetrovKP — CHANGES_REQUESTED** [review]

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i (hunk 1 above)
Comment (masdevas) [reply cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:219]: I decided use ptrs with checks

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 5]
```diff
@@ -573,19 +603,31 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::compute(N
 
         DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, n, sizeof(algorithmFPType));
 
-        TArrayScalable<algorithmFPType, cpu> aX(n * p);
-        TArrayScalable<algorithmFPType, cpu> aY(n);
+        if (_aX == nullptr || _aX->size() != n * p)
+        {
+            _aX = new TArrayScalable<algorithmFPType, cpu>(n * p);
```
Comment (owerbat) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:608]: If `_aX != nullptr && _aX->size() != n * p` then memory leak happens. You need `_aX->resize()` or something like that. The same in Log Loss

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 6]
```diff
@@ -573,19 +586,28 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::compute(N
 
         DAAL_OVERFLOW_CHECK_BY_MULTIPLICATION(size_t, n, sizeof(algorithmFPType));
 
-        TArrayScalable<algorithmFPType, cpu> aX(n * p);
-        TArrayScalable<algorithmFPType, cpu> aY(n);
-        s |= objective_function::internal::getXY<algorithmFPType, cpu>(dataNT, dependentVariablesNT, ntInd, aX.get(), aY.get(), nRows, n, p);
-        s |= doCompute(aX.get(), aY.get(), nRows, n, p, betaNT, valueNT, hessianNT, gradientNT, nonSmoothTermValue, proximalProjection,
-                       lipschitzConstant, parameter);
+        if (_aX.size() != n * p)
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:589]: If less than n*p

File: cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i [hunk 7]
```diff
@@ -573,19 +586,28 @@ services::Status CrossEntropyLossKernel<algorithmFPType, method, cpu>::compute(N
...
+        {
+            _aX.reset(n * p);
+            DAAL_CHECK_MALLOC(_aX.get());
+        }
+        if (_aY.size() != n)
```
Comment (PetrovKP) [inline cpp/daal/src/algorithms/objective_function/cross_entropy_loss/cross_entropy_loss_dense_default_batch_impl.i:594]: If less than n

**Review by PetrovKP — APPROVED** [review]
